spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java

```diff
@@ -252,22 +271,31 @@ else if (this.tree.isLimiting()) {
 			limit = this.tree.getMaxResults();
 		}
 
+		if (!singularResult && !total && pageable.isPaged()) {
+			limit = pageable.getPageSize();
+		}
+
 		Sort sort = this.tree.getSort();
 		if (getQueryMethod().getParameters().hasPageableParameter()) {
-			sort = sort.and(paramAccessor.getPageable().getSort());
+			sort = sort.and(pageable.getSort());
 		}
 		if (getQueryMethod().getParameters().hasSortParameter()) {
 			sort = sort.and(paramAccessor.getSort());
 		}
 
-		if (paramAccessor.getPageable().isPaged() && !total) {
-			//for slice queries we retrieve one additional item to check if the next slice exists
-			limit = paramAccessor.getPageable().getPageSize() + (isSliceQuery() ? 1 : 0);
-			offset = (int) paramAccessor.getPageable().getOffset();
+		if (pageable.isPaged() && !total) {
+			offset = (int) pageable.getOffset();
```

Comment:
Forgive me if I'm confused and have it all wrong here: 

Is the purposde of the cursor so that we can jump efficiently to page 54 of page size 10 of some query's result set right ? Pageable's getOffset method does that multiplication for you so it'll  do 54 * 10 = 540 offset , but if we already have the cursor wouldn't that be doing it twice? 
(I see here we apply both the offset that takes into account the page number as well as applying the cursor which means we will be starting on that correct page already? )

Reply:
Not exactly. When we create the actual query we only add offset when cursor is not available.